bootstrap/api/v1alpha1/rke2config_types.go

```diff
@@ -18,47 +18,624 @@ package v1alpha1
+	// +optional
+	LbServerPort string `json:"lbServerPort,omitempty"`
+
+	// Cni CNI Plugins to deploy, one of none, calico, canal, cilium; optionally with multus as the first value to enable the multus meta-plugin (default: canal)
+	// +optional
+	Cni Cni `json:"cni,omitempty"`
+
+	// PauseImage Override image to use for pause
+	// +optional
+	PauseImage string `json:"pauseImage,omitempty"`
+
+	// RuntimeImage Override image to use for runtime binaries (containerd, kubectl, crictl, etc)
+	// +optional
+	RuntimeImage string `json:"runtimeImage,omitempty"`
+
+	// KubeletPath Override kubelet binary path
+	// +optional
+	KubeletPath string `json:"kubeletPath,omitempty"`
+
+	//	CloudProviderName  Cloud provider name
+	// +optional
+	CloudProviderName string `json:"cloudProviderName,omitempty"`
+
+	//	CloudProviderConfigMap  is a reference to a ConfigMap containing Cloud provider configuration file path
+	// +optional
+	CloudProviderConfigMap ConfigMapRef `json:"cloudProviderConfigMap,omitempty"`
+
+	// NOTE: this was only profile, changed it to cisProfile
+	// CisProfile Validate system configuration against the selected benchmark (valid items: cis-1.23 )
+	// +optional
+	CisProfile CisProfile `json:"cisProfile,omitempty"`
+
+	// AuditPolicySecret Path to the file that defines the audit policy configuration
+	// +optional
+	AuditPolicySecret SecretRef `json:"auditPolicySecret,omitempty"`
+
+	// ControlPlaneResourceRequests Control Plane resource requests
+	// +optional
+	ControlPlaneResourceRequests string `json:"controlPlaneResourceRequests,omitempty"`
+
+	// ControlPlaneResourceLimits Control Plane resource limits
+	// +optional
+	ControlPlaneResourceLimits string `json:"controlPlaneResourceLimits,omitempty"`
+
+	// Etcd defines optional custom configuration of ETCD
+	//+optional
+	Etcd ComponentConfig `json:"etcd,omitempty"`
+
+	// KubeApiServer defines optional custom configuration of the Kube API Server
+	//+optional
+	KubeApiServer ComponentConfig `json:"kubeApiServer,omitempty"`
+
+	// KubeControllerManager defines optional custom configuration of the Kube Controller Manager
+	//+optional
+	KubeControllerManager ComponentConfig `json:"kubeControllerManager,omitempty"`
+
+	// KubeScheduler defines optional custom configuration of the Kube Scheduler
+	//+optional
+	KubeScheduler ComponentConfig `json:"kubeScheduler,omitempty"`
+
+	// KubeProxy defines optional custom configuration of the Kube Proxy
+	//+optional
+	KubeProxy ComponentConfig `json:"kubeProxy"`
+
+	// CloudControllerManager defines optional custom configuration of the Cloud Controller Manager
+	//+optional
+	CloudControllerManager ComponentConfig `json:"cloudControllerManager,omitempty"`
+
+	// Kubelet defines optional custom configuration of the Kubelet
+	//+optional
+	Kubelet ComponentConfig `json:"kubelet,omitempty"`
+}
+
+type RKE2AgentConfig struct {
+	// DataDir Folder to hold state
```

Comment:
I see a lot of duplications from the server config, what if we make a structure with a common field and have it nested in the server and client config?

Reply:
Yes, I agree, but what about "nested anonymous struct" where struct fields are simply promoted to the upper struct, like the following:

```go
type RKE2ServerConfig struct {
        // Same attributes as Agent
        RKE2AgentConfig

        // BindAddress describes the rke2 bind address (default: 0.0.0.0).
	//+optional
	BindAddress string `json:"bindAddress,omitempty"`

	// AdvertiseAddress IP address that apiserver uses to advertise to members of the cluster (default: node-external-ip/node-ip).
	//+optional
	AdvertiseAddress string `json:"advertiseAddress,omitempty"`

	// TLSSan Add additional hostname or IP as a Subject Alternative Name in the TLS cert.
	//+optional
	TLSSan []string `json:"tlsSan,omitempty"`
(...)
```
We could also just suppose that for the specific case of CAPI, the user must define the same common settings for Agent and Server, meaning, we completely remove the Agent settings from the Server.

What do you think?

Reply:
Acted to use "nested anonymous struct"